### PR TITLE
MAILPOET-4371 remove old model usage from post notification scheduler and replace it with doctrine

### PR DESCRIPTION
**Pull request by samnajian**
[MAILPOET-4371]

[MAILPOET-4371]: https://mailpoet.atlassian.net/browse/MAILPOET-4371?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: mailpoet/lib/Entities/ScheduledTaskEntity.php
Comment (lysyjan) [inline mailpoet/lib/Entities/ScheduledTaskEntity.php]: We declare task type constants in workers. It was discussed https://github.com/mailpoet/mailpoet/pull/4305#discussion_r942309426  in another PR recently. Also, the sending type is not the default.

File: mailpoet/lib/Newsletter/Scheduler/Scheduler.php
Comment (lysyjan) [inline mailpoet/lib/Newsletter/Scheduler/Scheduler.php:84]: Is this change necessary? I see that you created it probably due to storing `scheduledAt`. But maybe it would be easier to use `Carbon::createFromFormat` because we know the format. What do you think? You could also avoid formatting the date into logs.

File: mailpoet/lib/Newsletter/Scheduler/PostNotificationScheduler.php
Comment (lysyjan) [inline mailpoet/lib/Newsletter/Scheduler/PostNotificationScheduler.php]: I guess this annotation is redundant. In my opinion, the return type and parameter type are enough.

File: mailpoet/tests/integration/Newsletter/Scheduler/PostNotificationTest.php
Comment (lysyjan) [inline mailpoet/tests/integration/Newsletter/Scheduler/PostNotificationTest.php]: We want to remove all asserts from our codebase. There already exists [a ticket](https://mailpoet.atlassian.net/browse/MAILPOET-4258) for this.
If I'm not wrong, you can use `$this->assertInstanceOf()` to avoid errors from PHPStan.

You don't have to change the old code, but we shouldn't add more of those cases.

File: mailpoet/lib/Newsletter/Scheduler/PostNotificationScheduler.php
Comment (lysyjan) [inline mailpoet/lib/Newsletter/Scheduler/PostNotificationScheduler.php]: This method has name `createPostNotificationSendingTask` and you return `SendingQueueEntity`. I would expect to return `ScheduledTaskEntity` instead. Or is there a reason for this? I checked it, and we use the returned value only in tests.

File: mailpoet/tests/integration/Newsletter/Scheduler/PostNotificationTest.php
Comment (lysyjan) [inline mailpoet/tests/integration/Newsletter/Scheduler/PostNotificationTest.php]: I think that this check shouldn't be behind a condition. Please, don't forget to check the rest of your changes.

File: mailpoet/tests/integration/Newsletter/Scheduler/PostNotificationTest.php
Comment (lysyjan) [inline mailpoet/tests/integration/Newsletter/Scheduler/PostNotificationTest.php]: You can use [ScheduledTask factory](https://github.com/mailpoet/mailpoet/blob/trunk/mailpoet/tests/DataFactories/ScheduledTask.php) here.

File: mailpoet/tests/integration/Newsletter/Scheduler/PostNotificationTest.php
Comment (lysyjan) [inline mailpoet/tests/integration/Newsletter/Scheduler/PostNotificationTest.php]: This flush is redundant.

File: mailpoet/tests/integration/Newsletter/Scheduler/PostNotificationTest.php
Comment (lysyjan) [inline mailpoet/tests/integration/Newsletter/Scheduler/PostNotificationTest.php]: I have created a SendingQueue factory in [another PR](https://github.com/mailpoet/mailpoet/pull/4305/commits/a0cf7a817de934f720a3c907f0c7848fd62dbb1a). Unfortunately, it is not merged yet.

**Review by lysyjan — CHANGES_REQUESTED** [review]
I have a few remarks about your changes.

File: mailpoet/lib/Newsletter/Scheduler/PostNotificationScheduler.php
Comment (samnajian) [reply mailpoet/lib/Newsletter/Scheduler/PostNotificationScheduler.php]: The reason is that we have `$sendingQueue->getTask()` but not a method to get the queue from `ScheduledTaskEntity` and the returned values are only used in test as you mentioned, still I'd agree the name doesn't match much, will adapt

Comment (lysyjan) [reply mailpoet/lib/Newsletter/Scheduler/PostNotificationScheduler.php]: @samnajian It's possible to change the unidirectional relation to bidirectional. I did it in a different PR in this commit https://github.com/mailpoet/mailpoet/pull/4305/commits/33fe72bb5d1dcef0a1eb9ed1167cb6fc3543a09b.
You can read how it works [here](https://www.doctrine-project.org/projects/doctrine-orm/en/latest/reference/association-mapping.html#one-to-one-bidirectional).

You changed the method's name in the newest version, but other schedulers have the suffix `SendingTask`. So I think that we should preserve this pattern.

File: mailpoet/lib/Newsletter/Scheduler/Scheduler.php
Comment (samnajian) [reply mailpoet/lib/Newsletter/Scheduler/Scheduler.php:84]: I'd say this is necessary since I'd normally expect a higher lower level class to give me `raw`  information and not formatted one, and besides why should we format and create a string out of `DateTime` and then use the string to create `DateTime` again? It doesn't make much sense even though it's pretty easy to do so.

File: mailpoet/lib/Newsletter/Scheduler/PostNotificationScheduler.php
Comment (samnajian) [reply mailpoet/lib/Newsletter/Scheduler/PostNotificationScheduler.php]: Since you've added the code needed here in [this PR](https://github.com/mailpoet/mailpoet/pull/4305) it makes sense to either wait until the mentioned PR is merged or based [PR 4314](https://github.com/mailpoet/mailpoet/pull/4314) on it, what do you think?

Comment (lysyjan) [reply mailpoet/lib/Newsletter/Scheduler/PostNotificationScheduler.php]: @samnajian Yes, it makes sense.

File: mailpoet/lib/Newsletter/Scheduler/PostNotificationScheduler.php
Comment (lysyjan) [inline mailpoet/lib/Newsletter/Scheduler/PostNotificationScheduler.php]: I guess this flush is redundant because the sending queue is the owner of the relation.

File: mailpoet/lib/Newsletter/Scheduler/PostNotificationScheduler.php
Comment (lysyjan) [inline mailpoet/lib/Newsletter/Scheduler/PostNotificationScheduler.php]: This is a detail only, but the variable should be named `scheduledTask`. I'm fine with that, this is not a blocker.

**Review by lysyjan — COMMENTED** [review]
Only one small detail.

**Comment by lysyjan** [comment]
This PR should be merged after https://github.com/mailpoet/mailpoet/pull/4305

**Review by lysyjan — DISMISSED** [review]
Thank you for the changes and patience.

**Review by veljkho — APPROVED** [review]

**Pull request merged**
